substrate/frame/revive/src/vm/evm.rs

```diff
@@ -15,27 +15,55 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-mod instructions;
-
 use crate::{
+	exec::ExecError,
+	vec,
 	vm::{BytecodeType, ExecResult, Ext},
-	AccountIdOf, CodeInfo, Config, ContractBlob, DispatchError, Error, ExecReturnValue, H256,
-	LOG_TARGET,
+	AccountIdOf, Code, CodeInfo, Config, ContractBlob, DispatchError, Error, ExecReturnValue, H256,
+	LOG_TARGET, U256,
 };
-use alloc::vec::Vec;
+use alloc::{boxed::Box, vec::Vec};
+use core::cmp::min;
 use instructions::instruction_table;
 use pallet_revive_uapi::ReturnFlags;
 use revm::{
 	bytecode::Bytecode,
+	context::CreateScheme,
 	interpreter::{
 		host::DummyHost,
 		interpreter::{ExtBytecode, ReturnDataImpl, RuntimeFlags},
 		interpreter_action::InterpreterAction,
-		interpreter_types::InputsTr,
-		CallInput, Gas, Interpreter, InterpreterResult, InterpreterTypes, SharedMemory, Stack,
+		interpreter_types::{InputsTr, MemoryTr, ReturnData},
+		CallInput, CallInputs, CallScheme, CreateInputs, FrameInput, Gas, InstructionResult,
+		Interpreter, InterpreterResult, InterpreterTypes, SharedMemory, Stack,
 	},
-	primitives::{self, hardfork::SpecId, Address},
+	primitives::{self, hardfork::SpecId, Address, Bytes},
 };
+use sp_core::H160;
+use sp_runtime::Weight;
+
+mod instructions;
+
+/// Hard-coded value returned by the EVM `DIFFICULTY` opcode.
+///
+/// After Ethereum's Merge (Sept 2022), the `DIFFICULTY` opcode was redefined to return
+/// `prevrandao`, a randomness value from the beacon chain. In Substrate pallet-revive
+/// a fixed constant is returned instead for compatibility with contracts that still read this
+/// opcode. The value is aligned with the difficulty hardcoded for PVM contracts.
+pub(crate) const DIFFICULTY: u64 = 2500000000000000_u64;
+
+/// The base fee per gas used in the network as defined by EIP-1559.
+///
+/// For `pallet-revive`, this is hardcoded to 0
+pub(crate) const BASE_FEE: U256 = U256::zero();
+
+/// EVM max deployed runtime code size (EIP-170).
+/// 24,576 bytes (0x6000).
+pub const EVM_BYTECODE_LIMIT: usize = 24_576;
+
+/// EVM max initcode size (EIP-3860).
+/// 49,152 bytes (0xC000).
+pub const EVM_INITCODE_LIMIT: usize = EVM_BYTECODE_LIMIT * 2;
```

Comment:
these constants are defined already by 
revm::primitives::eip3860::MAX_INITCODE_SIZE 
and 
revm::primitives::eip170::MAX_CODE_SIZE 

let's not redefine them, these const are used below in from_evm_runtime / from_init_runtime

Reply:
done